[ci-visibility] Add manual testing API for javascript (#18773)
DataDog/documentation: content/en/continuous_integration/tests/javascript.md

From `dd-trace>=3.15.0` and `dd-trace>=2.28.0`, CI Visibility automatically uploads git metadata information (commit history). This metadata contains file names but no file contents. If you want to opt out of this behavior, you can do so by setting `DD_CIVISIBILITY_GIT_UPLOAD_ENABLED` to `false`. However, this is not recommended, as features like Intelligent Test Runner and others do not work without it.

## Manual testing API

<div class="alert alert-warning">

>   <strong>Note:</strong> To use the manual testing API, you must pass <code>DD_CIVISIBILITY_MANUAL_API_ENABLED=1</code> as an environment variable.

Reviewer: ```suggestion
  <strong>Note</strong>: To use the manual testing API, you must pass <code>DD_CIVISIBILITY_MANUAL_API_ENABLED=1</code> as an environment variable.
```

</div>

<div class="alert alert-warning">

>   <strong>Note:</strong> Manual testing API is currently in <strong>beta</strong>, so its API might change. It is is available from <code>dd-trace>=4.4.0</code>, <code>dd-trace>=3.25.0</code> and <code>dd-trace>=2.38.0</code>.

Reviewer: ```suggestion
  <strong>Note</strong>: The manual testing API is in <strong>beta</strong>, so its API might change. It is available starting in <code>dd-trace</code> versions <code>4.4.0</code>, <code>3.25.0</code>, and <code>2.38.0</code>.
```

</div>

> If you use Jest, Mocha, Cypress, Playwright or Cucumber, CI Visibility automatically instruments them and sends the test results to Datadog. If you use an unsupported testing framework or if you have a different testing mechanism, you can instead use the API to report test results to Datadog.

Reviewer: ```suggestion
If you use Jest, Mocha, Cypress, Playwright, or Cucumber, CI Visibility automatically instruments them and sends the test results to Datadog. If you use an unsupported testing framework or if you have a different testing mechanism, you can instead use the API to report test results to Datadog.
```

The manual testing API leverages the `node:diagnostics_channel` module from Node.js and is based on channels you can publish to:

```javascript
const { channel } = require('node:diagnostics_channel')

const { describe, test, beforeEach, afterEach, assert } = require('my-custom-test-framework')

const testStartCh = channel('dd-trace:ci:manual:test:start')
const testFinishCh = channel('dd-trace:ci:manual:test:finish')
const testSuite = __filename

describe('can run tests', () => {
  beforeEach((testName) => {
    testStartCh.publish({ testName, testSuite })
  })
  afterEach((status, error) => {
    testFinishCh.publish({ status, error })
  })
  test('first test will pass', () => {
    assert.equal(1, 1)
  })
})
```

### Test start channel

> Grab this channel by its ID `dd-trace:ci:manual:test:start` to publish that a test is going to start. A good place to do this might be a `beforeEach` hook, or similar.

Reviewer: ```suggestion
Grab this channel by its ID `dd-trace:ci:manual:test:start` to publish that a test is starting. A good place to do this is a `beforeEach` hook or similar.
```

```typescript
const { channel } = require('node:diagnostics_channel')
const testStartCh = channel('dd-trace:ci:manual:test:start')

// ... code for your testing framework goes here
  beforeEach(() => {
    const testDefinition = {
      testName: 'a-string-that-identifies-this-test',
      testSuite: 'what-suite-this-test-is-from.js'
    }
    testStartCh.publish(testDefinition)
  })
// code for your testing framework continues here ...
```

> The payload that needs to be published has attributes `testName` and `testSuite`, both of which are strings, that identify the test about to start.

Reviewer: ```suggestion
The payload to be published has attributes `testName` and `testSuite`, both strings, that identify the test that is about to start.
```

### Test finish channel

> Grab this channel by its ID `dd-trace:ci:manual:test:finish` to publish that a test is about to end. A good place to do this might be a `afterEach` hook, or similar.

Reviewer: ```suggestion
Grab this channel by its ID `dd-trace:ci:manual:test:finish` to publish that a test is ending. A good place to do this is an `afterEach` hook or similar.
```

```typescript
const { channel } = require('node:diagnostics_channel')
const testFinishCh = channel('dd-trace:ci:manual:test:finish')

// ... code for your testing framework goes here
  afterEach(() => {
    const testStatusPayload = {
      status: 'fail',
      error: new Error('assertion error')
    }
    testStartCh.publish(testStatusPayload)
  })
// code for your testing framework continues here ...
```

> The payload that needs to be published has attributes `status` and `error`:

Reviewer: ```suggestion
The payload to be published has attributes `status` and `error`:
```

> * `status` is a string that can only take three values:

Reviewer: ```suggestion
* `status` is a string that takes one of three values:
```

  * `'pass'` when a test passes.
  * `'fail'` when a test fails.
  * `'skip'` when a test has been skipped.

* `error` is an `Error` object containing the reason why a test failed.

### Add tags channel

Grab this channel by its ID `dd-trace:ci:manual:test:addTags` to publish that a test needs custom tags. This can be done within the test function:

```typescript
const { channel } = require('node:diagnostics_channel')
const testAddTagsCh = channel('dd-trace:ci:manual:test:addTags')

// ... code for your testing framework goes here
  test('can sum', () => {
    testAddTagsCh.publish({ 'test.owner': 'my-team', 'number.assertions': 3 })
    const result = sum(2, 1)
    assert.equal(result, 3)
  })
// code for your testing framework continues here ...
```

> The payload that needs to be published is a dictionary `<string, string|number>` of tags or metrics that will be added to the test.

Reviewer: ```suggestion
The payload to be published is a dictionary `<string, string|number>` of tags or metrics that are added to the test.
```

> ### How to run

Reviewer: ```suggestion
### Run the tests
```

> Once the test start and end channels are put in place in your code, you may run your testing framework like you normally do, including the following environment variables:

Reviewer: ```suggestion
When the test start and end channels in your code, run your testing framework like you normally do, including the following environment variables:
```

Author: would this be 
```
When the test start and end channels **are** in your code, run your testing...
```
?

Reviewer: Yes, **are** :)

Author: thanks!
